config/configgrpc/configgrpc_test.go

```diff
@@ -71,7 +72,7 @@ func TestAllGrpcClientSettings(t *testing.T) {
 		{
 			name: "test all with gzip compression",
 			settings: GRPCClientSettings{
-				Headers: map[string]string{
+				Headers: configopaque.Headers{
```

Comment:
It's possible, but painful. We could do it as follows (focusing on the example you mentioned, similar process for other structs):
1. (Release v0.N) Add a new `OpaqueHeaders` field with the new type and deprecate `Headers` in favor of it. Modify code to take into account both fields (and log a warning maybe?)
3. (Release v0.N+1) Change `Headers` field to have opaque type and remove deprecation, deprecate `OpaqueHeaders` field in favor of `Headers`
4. (Release v0.N+2) Remove `OpaqueHeaders` field.

If someone is not using the methods on the structs to build the client/server they would have to write custom code during the migration. One could argue this is a misuse of the API and we should not care about this particular case.

Do you think it's worth doing this? I think it's important to settle on what we want to do here, since we also have to follow a similar process then for every component on contrib that has sensitive fields

Reply:
> If someone is not using the methods on the structs to build the client/server they would have to write custom code during the migration

Sorry, I do not feel like I understood what you meant here. People using the `Headers` struct field will have to use the new `opaque.String` sooner or later, right? Sooner means breaking them now, forcing them to migrate right away, potentially at a critical time for them. Later means deprecating the current Header, pointing people to OpaqueHeader, and then moving back to Header (now backed by opaque.String). 

It might be a lot of work, but I think doing it the way we specified earlier for breaking changes should be followed unless there's a good reason not to do it (technical unfeasibility being a good one, which doesn't seem to be the case here).

Reply:
> Sorry, I do not feel like I understood what you meant here

I meant that if someone is using `GRPCClientSettings` but not building the client through [`ToClientConn`](https://pkg.go.dev/go.opentelemetry.io/collector/config/configgrpc#GRPCClientSettings.ToClientConn) we can't support them during the transition to opaque type: they are on their own when it comes to handling both `OpaqueHeaders` and `Headers`. I think that's fine, but wanted to point it out explicitly.

> It might be a lot of work, but I think doing it the way we specified earlier for breaking changes should be followed unless there's a good reason not to do it (technical unfeasibility being a good one, which doesn't seem to be the case here).

No technical reason, I just wanted to make sure we want to commit to following this process for the 30+ fields we will have to change on contrib. Sounds like this is the only way to go, so I will keep this PR with the new package and split the OpaqueHeaders/Headers to separate PRs

Reply:
> As bad as it sounds, I think this is one of the positive side effects of having the process: it requires us to evaluate carefully if breaking our users (ourselves being our first users) is worth the change.

I think the process is useful in that sense if there are actually users of the thing you are breaking, I have personally yet to see a Go API user who is accessing individual fields on a contrib component config struct :) 

But regardless, I get that I am the minority view here, and that it is technically possible to do this in several steps with the same end result, so let's go for that.